feat: add kubescape patch command

anubhav06 opened the pull request:
## Overview
Add `kubescape patch` command

## Additional Information
1. The `kubescape patch` command can be used to patch container images with vulnerabilities.
2. It uses [copa](https://github.com/project-copacetic/copacetic) and [buildkit](https://github.com/moby/buildkit) under the hood for patching the images, and [grype](https://github.com/anchore/grype) as its engine for scanning the images (at the moment)
3. The detailed documentation for this command can be found [here](https://github.com/anubhav06/kubescape/tree/patch-cmd/cmd/patch#readme).

## TODO

1. Replace `anubhav06/copacetic` with `project-copacetic/copacetic`, when the copa team accepts the kubescape and copa integration support PR.

## Usage

```bash
sudo buildkitd & 
sudo kubescape patch --image <image-name>
```
> The patch command can also be run without sudo privileges. Refer to the documentation [here](https://github.com/anubhav06/kubescape/tree/patch-cmd/cmd/patch#readme).

## Examples/Screenshots
1. Run `sudo buildkitd` in a separate terminal  

![0](https://github.com/kubescape/kubescape/assets/43821031/43ef7cfa-6155-481e-af40-0a59c44ab83d)

2. Run `sudo kubescape patch --image docker.io/library/nginx:1.22`:  

![image](https://github.com/kubescape/kubescape/assets/43821031/a74e68e9-d99b-4c85-b76d-6a31054faf07)
![image](https://github.com/kubescape/kubescape/assets/43821031/a2077139-8901-4bf9-b172-8d892804d666)

cc @craigbox @dwertent 
## Related issues/PRs:

Resolved #1227 



## Checklist before requesting a review

put an [x] in the box to get it checked 

- [x] My code follows the style guidelines of this project
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.
- [x] New and existing unit tests pass locally with my changes

craigbox commented:
At the present, there's no assertion of levels of support in Kubescape.  Everything just "is", in the most part.
I have proposed that we introduce [feature gates](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) (with the same syntax as the Kubernetes API server) in order to be able to only include Alpha or Labs-level features if someone enables a feature gate. 

The alternative would be keeping it in a separate branch, which would require keeping it in sync, and making builds from it if we wanted anyone to actually use it.

/cc @rotemamsa

anubhav06 commented:
The former sounds better to me. So this will be merged later, when the feature gates have been introduced, right?

anubhav06 commented:
> Also, you will see the CLI output, it is essential we keep it aligned.  
1. Ok, so the vulnerability summary that we see in the CLI output of patch command, it should be same as that of the image scan command? Right now, for the patch command I've modified the output to show the Total CVE's and Fixable CVE's. I should remove this and format it to look the same way as we see in the image scan output, right?

> (and there is no reason to develop another "image scanning" mechanism in Kubescape)
2. I'm unable to understand this. For the patch command. I've just replicated the image scanning mechanism of `kubescape scan image` command. Similar replication is also used while running `kubescape scan --scan-images` as shown [here](https://github.com/kubescape/kubescape/blob/master/core/core/scan.go#L255-L281).

anubhav06 commented:
@dwertent I've updated the output to align with that of image scan command ([screenshot in PR desc](https://github.com/kubescape/kubescape/pull/1332#issue-1837226194)). 
I've also removed the "image scanning" logic I'd added for results handling.
Please let me know if this is what's expected and if any other changes are required.

Daniel-GrunbergerCA commented:
@anubhav06 This feature looks really cool and works smoothly! I'm very excited for this!

I have some general comments about the feature, which are mostly food for thought:

1. We should add the full canonical name to the `kubescape patch` follow-up step (which is printed at the end). This would be the right call to action, especially since the `kubescape scan image` command works without canonical name, and it might confuse the user
2. Maybe we should add to the output some comparison between the old image and the new one, so it will become clear to the user the value he is getting from the `patch` command
3. Do we need to print the entire copa output? I think is best to have it in debug mode only
4. We shouldn't use the `-r` flag. We have the `-f` flag, where we can specify JSON output

cc: @craigbox @dwertent

Daniel-GrunbergerCA commented on core/core/patch.go:
```diff
@@ -0,0 +1,113 @@
+	fileNamePatched := fmt.Sprintf("%s:%s.json", patchInfo.ImageName, patchInfo.PatchedImageTag)
+	fileNamePatched = strings.ReplaceAll(fileNamePatched, "/", "-")
+
+	if patchInfo.IncludeReport {
```
this logic should be handled by the `HandleResults` function. Besides that, I don't see why we need a new flag, instead of using the `-f` for choosing the format

Daniel-GrunbergerCA replied:
yes, thank you

Daniel-GrunbergerCA commented on cmd/patch/patch.go:
```diff
@@ -0,0 +1,94 @@
+	patchCmd.PersistentFlags().StringVarP(&patchInfo.PatchedImageTag, "tag", "t", "", "Tag for the patched image. Defaults to '<image-tag>-patched' ")
+	patchCmd.PersistentFlags().StringVarP(&patchInfo.BuildkitAddress, "address", "a", "unix:///run/buildkit/buildkitd.sock", "Address of buildkitd service, defaults to local buildkitd.sock")
+	patchCmd.PersistentFlags().DurationVar(&patchInfo.Timeout, "timeout", 5*time.Minute, "Timeout for the operation, defaults to '5m'")
+	patchCmd.PersistentFlags().BoolVarP(&patchInfo.IncludeReport, "report", "r", false, "Generate & save a before and after report for the image (json format))")
```
is there a reason for not using `--format` as we have in other commands?

Daniel-GrunbergerCA commented on core/pkg/resultshandling/printer/v2/jsonprinter.go:
```diff
@@ -75,8 +75,8 @@ func printConfigurationsScanning(opaSessionObj *cautils.OPASessionObj, ctx conte
 func (jp *JsonPrinter) PrintImageScan(ctx context.Context, scanResults *models.PresenterConfig) error {
-	presenterConfig, _ := presenter.ValidatedConfig("json", "", false)
-	pres := presenter.GetPresenter(presenterConfig, *scanResults)
+	// presenterConfig, _ := presenter.ValidatedConfig("json", "", false)
```
you can remove it altogether

Daniel-GrunbergerCA commented on cmd/patch/patch.go:
```diff
@@ -0,0 +1,94 @@
+
+	// Check if image is in canonical format (required by copacetic for patching images)
+	// Parse the image full name to get image name and tag
+	named, err := ref.ParseNamed(patchInfo.Image)
```
let's abstract way this detail from the user. Instead of asking him for the canonical name, let's accept an image name as we do in the `scan image` command (such as "nginx") and build the canonical name ourselves.
You can use the `normalize_image_name` inside `opaprocessor`

anubhav06 replied:
since `normalize_image_name` lies under a different package, is it ok if I change it to an exported function? doing so, will allow me to access the function

Daniel-GrunbergerCA replied:
Let's change it so you can export the function. This way we will handle this logic only on one place

dwertent replied:
Yes, you could "move it" to the cautils package

Daniel-GrunbergerCA commented on core/pkg/resultshandling/printer/v2/prettyprinter/utils.go:
```diff
@@ -27,6 +27,7 @@ var (
 	clusterScanRunText   = fmt.Sprintf("Run a cluster scan: %s", getCallToActionString("'$ kubescape scan'"))
 	installHelmText      = fmt.Sprintf("Install Kubescape in your cluster for continuous monitoring: %s", linkToHelm)
 	CICDSetupText        = fmt.Sprintf("Add Kubescape to your CI/CD pipeline: %s", linkToCICDSetup)
+	imagePatchText		 = fmt.Sprintf("Try patching your images with: %s", getCallToActionString("'$ kubescape patch'"))
```
please note that this text is being printed even when running the `patch` command itself

anubhav06 replied:
yes I noticed that, sorry about that, will fix this as well.

anubhav06 commented:
@Daniel-GrunbergerCA to answer your questions:
1. on it
2. Even I agree with that. I had done that only previously (screenshot [here](https://github-production-user-asset-6210df.s3.amazonaws.com/43821031/258508039-d35bb19c-e6c7-4387-b7d6-ff4a1cddd9f5.png)). But since, David said that the output should align with that of the image scan command, I removed that. 
3. Alright. So I'll try to remove the output of copa to STDOUT and display it only when running with debug mode, and see how it looks. (Personally I believe that it looks good right now, the way it is, because by default the user can see what's happening: copa is building the image, installing the pkg and exporting to docker). But yes, I'll try to do it the way you said and let's see how it looks.
4. on it

Daniel-GrunbergerCA commented:
> 
You can wait with that, let's see what @craigbox and @dwertent think

dwertent commented:
2. @Daniel-GrunbergerCA I think it is a good suggestion, but still, I believe we should print only the latest state so we don't overwhelm the terminal. Also, this way we can have the same function for printing the `image scan` command and `image patch` command. 
As I'm writing this reply, I'm thinking we can print just the summary of the old image, but we can save this to the future.
3. Thenk you @anubhav06 . The reason Daniel asked to "hide" the logs is because it does not look good when using two different loggers when running a single command. I suggest we should use the spinner (form the kubescape logger repo) instead of printing the copa logs. I'm aware that we are less "verbose", but for now I think it's better :)

In general, thank you @anubhav06 for this PR!
Please merge the changes that have been pushed to the main so we will be able to properly test your PR before accepting it :)

anubhav06 commented:
changes:
- By default, copa's output (logs) will not be shown. 
If a user wants to see copa's logs, run kubescape in debug mode ( `$KS_LOGGER_LEVEL='debug'`)
- removed the `--report` flag as it was redundant & was not providing any real value
- moved `normalize_image_name` from `opaprocessor` pkg to `cautils` pkg
- removed the "try patching your image with..." text (for now)

dwertent commented on core/pkg/opaprocessor/utils.go:
```diff
@@ -106,6 +106,6 @@ var imageNameNormalizeDefinition = func(bctx rego.BuiltinContext, a *ast.Term) (
 		return nil, fmt.Errorf("invalid parameter type: %v", err)
 	}
-	normalizedName, err := normalize_image_name(string(aStr))
+	normalizedName, err := cautils.Normalize_image_name(string(aStr))
```
@YiscahLevySilas1 can this affect the controls that are using this function?

YiscahLevySilas1 replied:
I don't think it will affect them

dwertent commented on core/pkg/resultshandling/printer/v2/sarifprinter.go:
```diff
@@ -116,12 +116,7 @@ func (sp *SARIFPrinter) printImageScan(scanResults *models.PresenterConfig) erro
 		return fmt.Errorf("no no image vulnerability data provided")
 	}
 
-	presenterConfig, err := presenter.ValidatedConfig(printer.SARIFFormat, "", false)
```
was this accidentally removed?
cc @Daniel-GrunbergerCA

anubhav06 replied:
No. Because with the new version of grype, the previous syntax is not supported. 
Similarly, in the `jsonprinter`, the `ValidatedConfig` method has been removed ([here](https://github.com/kubescape/kubescape/pull/1332/files#diff-6c52364e93973c10e45bf3a765f082f55e92004a744438374dd212f188c4ff8aR81))

Daniel-GrunbergerCA commented on core/core/patch.go:
```diff
@@ -0,0 +1,116 @@
+
+func (ks *Kubescape) Patch(ctx context.Context, patchInfo *ksmetav1.PatchInfo) error {
+
+	logger.InitLogger(iconlogger.LoggerName)
```
remove this line (it overrides logger initialization)

Daniel-GrunbergerCA commented on core/core/patch.go:
```diff
@@ -0,0 +1,116 @@
+		return err
+	}
+
+	if logger.L().GetLevel() != "debug" {
```
why do we need to check it twice?

anubhav06 replied:
removed this, added a single line now: `os.Stdout, os.Stderr = sout, serr` (to restore the output)
